src/core/lib/resolver/server_address.h

```diff
@@ -32,44 +31,74 @@
-  // uniqueness or behavior.  They are for use by LB policies only.
-  //
-  // Attributes are keyed by a C string that is unique by address, not
-  // by value.  All attributes added with the same key must be of the
-  // same type.
   class AttributeInterface {
    public:
     virtual ~AttributeInterface() = default;
 
+    // The type name for this attribute.
+    // There can be only one attribute of a given type in a given AttributeMap.
+    // Attributes are keyed by the address of the string, not the value.
+    virtual const char* type() const = 0;
+
     // Creates a copy of the attribute.
     virtual std::unique_ptr<AttributeInterface> Copy() const = 0;
 
     // Compares this attribute with another.
-    virtual int Cmp(const AttributeInterface* other) const = 0;
+    virtual int Compare(const AttributeInterface* other) const = 0;
 
     // Returns a human-readable representation of the attribute.
     virtual std::string ToString() const = 0;
   };
 
-  // Takes ownership of args.
-  ServerAddress(const grpc_resolved_address& address, grpc_channel_args* args,
-                std::map<const char*, std::unique_ptr<AttributeInterface>>
-                    attributes = {});
-  ServerAddress(const void* address, size_t address_len,
-                grpc_channel_args* args,
-                std::map<const char*, std::unique_ptr<AttributeInterface>>
-                    attributes = {});
+  ResolverAttributeMap() = default;
+
+  explicit ResolverAttributeMap(
+      std::vector<std::unique_ptr<AttributeInterface>> attributes);
+
+  // Copyable.
+  ResolverAttributeMap(const ResolverAttributeMap& other);
+  ResolverAttributeMap& operator=(const ResolverAttributeMap& other);
+
+  // Movable.
+  ResolverAttributeMap(ResolverAttributeMap&& other) noexcept;
+  ResolverAttributeMap& operator=(ResolverAttributeMap&& other) noexcept;
+
+  int Compare(const ResolverAttributeMap& other) const;
+
+  // Returns the attribute for the specified key, or null if not present.
+  const AttributeInterface* Get(const char* key) const;
+
+  // Adds attribute to the map.
+  void Set(std::unique_ptr<AttributeInterface> attribute);
+
+  // Removes the attribute from the map.
+  void Remove(const char* key);
+
+  std::string ToString() const;
+
+  bool empty() const { return map_.empty(); }
+
+ private:
+  std::map<const char*, std::unique_ptr<AttributeInterface>> map_;
+};
+
+//
+// ServerAddress
+//
+
+// A server address is an address with two sets of attributes attached,
+// one set that affects subchannel behavior, and another set that is
+// visible only to LB policies.
+// TODO(roth): Remove grpc_channel_args from this class.
+class ServerAddress {
+ public:
+  ServerAddress(
+      const grpc_resolved_address& address, grpc_channel_args* args,
+      ResolverAttributeMap subchannel_attributes = ResolverAttributeMap(),
+      ResolverAttributeMap lb_policy_attributes = ResolverAttributeMap());
```

Comment:
This is getting to the point where I'd want to suggest a `ServerAddressBuilder` to assist putting these things together.